File: cpp/74_搜索二维矩阵_medium.cpp
class Solution {
public:
    bool searchMatrix(vector<vector<int>>& matrix, int target) {
        int row = matrix.size();
        int col = matrix[0].size();
        int mid = 0;
        int top = 0;
        int bottom = row;
        while (top < bottom) {
            mid = (top+bottom) / 2;
            if (target == matrix[mid][0])
                return true;
            else if (target < matrix[mid][0])
                bottom--;
            else {
                if (target == matrix[mid][col-1])
                    return true;
                else if (target < matrix[mid][col-1])
                    return findx(matrix[mid], target);
                else
                    top++;
            }
        }
        return false;
    }
    bool findx(vector<int> &matrix_rows, int target) {
        for (auto num : matrix_rows) {
            if (target == num)
                return true;
        }
        return false;
    }
};

File: python/74_搜索二维矩阵_medium.py
class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        def findx(matrix_row):
            for num in matrix_row:
                if target == num:
                    return True
            return False

        row, col = len(matrix), len(matrix[0])
        top, bottom, mid = 0, row, 0
        while top < bottom:
            mid = (top + bottom) // 2
            if target == matrix[mid][0]:
                return True
            elif target < matrix[mid][0]:
                bottom -= 1
            else:
                if target == matrix[mid][-1]:
                    return True
                elif target < matrix[mid][-1]:
                    return findx(matrix[mid])
                else:
                    top += 1
        return False

